casper/src/main/scala/coop/rchain/casper/package.scala

```diff
@@ -10,7 +10,7 @@ package object casper {
   type BlockProcessing[A]   = Either[BlockError, A]
   type ValidBlockProcessing = BlockProcessing[ValidBlock]
 
-  type ProposeFunction[F[_]] = (Casper[F] => F[Option[Int]])
+  type ProposeFunction[F[_]] = ((Casper[F], Boolean) => F[Option[Either[Int, BlockHash]]])
```

Comment:
Using nested `Option[Either[_,_]]` and other combinations is very hard to understand. I'm ok with this for now but we will need better solution.